pkg/xds/generator/protocol.go

```diff
@@ -0,0 +1,63 @@
+package generator
+
+import (
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+	mesh_core "github.com/Kong/kuma/pkg/core/resources/apis/mesh"
+	core_xds "github.com/Kong/kuma/pkg/core/xds"
+)
+
+var (
+	// protocolStack is a mapping between a protocol and its full protocol stack, e.g.
+	// HTTP has a protocol stack [HTTP, TCP],
+	// GRPC has a protocol stack [GRPC, HTTP2, TCP],
+	// TCP  has a protocol stack [TCP].
+	protocolStacks = map[mesh_core.Protocol]mesh_core.ProtocolList{
+		mesh_core.ProtocolHTTP: mesh_core.ProtocolList{mesh_core.ProtocolHTTP, mesh_core.ProtocolTCP},
+		mesh_core.ProtocolTCP:  mesh_core.ProtocolList{mesh_core.ProtocolTCP},
+	}
+)
+
+// GetCommonProtocol returns a common protocol between given two.
+//
+// E.g.,
+// a common protocol between HTTP and HTTP  is HTTP,
+// a common protocol between HTTP and TCP   is TCP,
+// a common protocol between GRPC and HTTP2 is HTTP2,
+// a common protocol between HTTP and HTTP2 is TCP.
+func GetCommonProtocol(one, another mesh_core.Protocol) mesh_core.Protocol {
```

Comment:
It's public mostly for convenience in unit testing.

We don't write unit tests only because a function is public. We write unit tests when we want to test a function in isolation (because we see a value in doing it this way instead of testing indirectly).

Exporting a function makes unit testing simpler. However, it's not a pre-requirement.

I've renamed `GetCommonProtocol` into `getCommonProtocol` and updated unit test.

Reply:
I write tests to check functionality, not functions. At least that is what I did until I started writing in Go. The convention of putting tests next to the file kind of tries to force you to write tests to functions.

The problem I have with this approach is that when I want to change the functionality I need to now change tests of two functions. Refactor is harder because if I change anything in the code I need to change functions tests since they are 1:1. Tests should help me refactor code whenever I want, not make it harder.

> Exporting a function makes unit testing simpler. However, it's not a pre-requirement.

I don't think we should ever make function public just for the sake of simpler testing. Proper encapsulation makes my reasoning about code much better. I instantly know something is local  to this package. Once I see something is public I check where in the other parts of the system it is used.